packages/atclient/src/notification.c

```diff
@@ -0,0 +1,322 @@
+}
+
+void atclient_notify_params_free(atclient_notify_params *params) {
+  atclient_atkey_free(&params->key);
+  free(params->value);
+  free(params->notifier);
+}
+
+int atclient_notify(atclient *ctx, atclient_notify_params *params) {
+  int res = 1;
+  // Step 1 calculate the buffer size needed for the protocol command
+  size_t bufsize = 6 + 1 + 2 + 1 + 36 + 3 +              // "notify:id:" + $uuid + trailing '\r\n\0'
+                   1 + 8 + 1 + strlen(params->notifier); // ":notifier:$notifier"
+
+  // minimum viable: notify:update:key:value -> id
+  // TODO: check for bad input in params struct since we don't provide the best helper functions
+  // and add valid ones to bufsize
+
+  if (params->operation != NO_none) {
+    bufsize += 1 + 6; // ":update" | ":delete"
+  }
+
+  if (params->message_type != NMT_none) {
+    bufsize += 1 + 11 + 1 + 4; // ":messageType" + ":text" | ":key"
+  }
+
+  if (params->priority != NP_none) {
+    bufsize += 1 + 8 + 1 + 6; // ":priority" + ":low" | ":medium" | ":high"
+  }
+
+  if (params->strategy != NS_none) {
+    bufsize += 1 + 8 + 1 + 6; // ":strategy" + ":all" | ":latest"
+  }
+
+  if (params->notification_expiry > 0) {
+    bufsize += 1 + 4 + 1 + // ":ttln:"
+               20;         // epochMillis (20 digits covers all 2^64 of unsigned long long, good for 300,000+ years)
+  }
+
+  if (params->value != NULL) {
+    bufsize += 1 + strlen(params->value); // :$value
+  }
+
+  // add metadata fragment length
+  size_t metadatalen = atclient_atkey_metadata_protocol_strlen(&params->key.metadata);
+  bufsize += metadatalen;
+
+  // atkey parts length
+  size_t atkeylen = atclient_atkey_strlen(&params->key);
+  bufsize += 1 + atkeylen; // :$atkey
+
+  // Step 2 generate / retrieve values which could potentially fail
+  res = atchops_uuid_init();
+  if (res != 0) {
+    atclient_atlogger_log("atclient | notification", ATLOGGER_LOGGING_LEVEL_WARN,
+                          "atchops_uuid_init failed with code %d\n", res);
+    return res;
+  }
+
+  // careful about when this is called, since it will write a null terminator to the 37th char
+  res = atchops_uuid_generate(params->id, 37);
+  if (res != 0) {
+    atclient_atlogger_log("atclient | notification", ATLOGGER_LOGGING_LEVEL_WARN,
+                          "atchops_uuid_generate failed with code %d\n", res);
+    return res;
+  }
+
+  struct timeval tv;
+  res = gettimeofday(&tv, NULL);
+  if (res != 0) {
+    atclient_atlogger_log("atclient | notification", ATLOGGER_LOGGING_LEVEL_WARN,
+                          // TODO: get errno
+                          "gettimeofday failed with code stored in errno %d\n", res);
+    return res;
+  }
+
+  unsigned long long ttln =
+      (unsigned long long)(tv.tv_sec) * 1000 + (unsigned long long)(tv.tv_usec) / 1000 + params->notification_expiry;
+
+  // Step 3 allocate the buffer and populate the full command
+  char cmd[bufsize];
+  snprintf(cmd, bufsize, "notify:id:%s", params->id);
```

Comment:
finally create the command buffer and populate it with all the bits.

Reply:
Now that I am sitting here documenting this, I probably could of just written directly to the socket, instead of computing an intermediate buffer.